### ISSUE TITLE
Document how to freeze layers

### ISSUE THREAD

**Issue by ottonemo** [dnouri/skorch#261]
See #236.

We should add a minimal example of how to filter parameters with `requires_grad=False` from the optimizer. E.g. as proposed in https://github.com/dnouri/skorch/issues/236#issuecomment-392579800:

```python
def filtered_optimizer(pgroups, **kwargs):
    params = filter(lambda t: t[1].requires_grad, pgroups[0]['params'])
    return torch.optim.SGD(params, **kwargs)

net = Net(module, optimizer=filtered_optimizer)
```

This depends on #260 which would help generalize this example by supporting lists and param group dictionaries alike.

**Comment by thomasjpfan** [comment]
The pytorch `Optimizer` class only checks `require_grad` during initialization. Thus something like this will work:

```python
module = module_cls()
net = net_cls(module)

net.initialize()
# Freeze first layer
for p in module.dense0.parameters():
    p.requires_grad_(False)

X, y = data
net.partial_fit(X, y)

# Unfreeze first layer
for p in module.dense0.parameters():
    p.requires_grad_(True)

net.partial_fit(X, y)
```

During the first `net.initalize()` all of the module's parameters have set `require_grad==True`, thus `net.optimizer_` is created without raising an exception. Then we freeze the first layer, and call `paritial_fit` to avoid initializing again. Next, we unfreeze the layer and call `partial_fit`. This setup will only fail when you freeze all the layers in the module. 

This works because `parameter.grad` is `None` when `requires_grad==False`. For example:

```python
x = torch.ones(2, 2, requires_grad=True)
y = x + 2
z = y * y * 3
out = z.mean()
out.backward()

assert x.grad is not None
assert y.grad is None
```

All optimizers checks for `parameter.grad` first before updating any parameters.

**Comment by benjamin-work** [comment]
> The pytorch Optimizer class only checks require_grad during initialization

I would assume that most of the times when someone wants to freeze layers, they have a pretrained net ready and only want to update parts of it (e.g. imagenet transfer learning). In that case, the module would already be initialized and skorch would not reset the module. So that should work out fine.

In case I have an uninitialized module, I would not to initialize it first, which sucks. However, if we have a solution for #107, it should be possible to make it work with uninitialized modules by setting `net = NeuralNet(..., module__weight0__requires_grad=False)`.

**Comment by thomasjpfan** [comment]
I am trying to come up with "the recommend way" to do fine tuning in skorch. This entails, having layers that are frozen for some epochs, and then unfreezing some layers for additional training.

If there does exist a `filtered_optimizer` method. Fine tuning would go like this:

```python
# some layers are already frozen
module = module_cls()
net = Net(module, optimizer=filtered_optimizer)

X, y = data
net.fit(X, y)

# Unfreeze some layer
for p in module.dense0.parameters():
    p.requires_grad_(True)

net.initialize_optimizer()
net.partial_fit(X, y)
```

`initialize_optimizer` needs to be called to update the `parameter_groups` in `optimizer_` to reflect the changes in `requires_grad`.

I think the this can all be simplified, by creating `optimizer_` in the beginning of `fit_loop`, where, by default, the `require_grads==False` are filtered out. This way a user not need to pass in `filtered_optimizer`. This way fine tuning would look like:

```python
# some layers are already frozen
module = module_cls()
net = Net(module, optimizer=torch.optim.SGD)

X, y = data
net.fit(X, y)

# Unfreeze some layer
for p in module.dense0.parameters():
    p.requires_grad_(True)

net.partial_fit(X, y)
```

I would prefer an API that looks like this.

**Comment by ottonemo** [comment]
If I understand you correctly, your argument is that `net.initialize_optimizer()` should be implicit in each `fit_loop` call to avoid a line of code on the user side. While I think that it is important to minimize overhead on the user's side I don't think that it is beneficial in this case. Firstly, by moving the call from the general initialization method to `fit_loop` we break up a very clear structure which needs to be justified and explained somewhere (adds maintenance cost and code complexity). Secondly, this change actually limits the user in their expressiveness. In the current state it is quite possible to re-initialize the optimizer after calling `partial_fit` several times while preserving the state. When moving the optimizer initialization to `fit_loop` we explicitly prevent the user from keeping the optimizer's state for several fits.

**Comment by thomasjpfan** [comment]
I see your point. A user would lose more control if `net.initialize_optimizer()` was in `fit_loop`. 


Considering comments in #262 and here, I propose this for fine tuning:

```python
from skorch.helper import filtered_optimizer

# some layers are already frozen
module = module_cls()
net = Net(module, optimizer=filtered_optimizer(torch.optim.SGD))

X, y = data
net.fit(X, y)

# Unfreeze some layer
for p in module.dense0.parameters():
    p.requires_grad_(True)

net.initialize_optimizer()
net.partial_fit(X, y)
```

What do you think?

**Comment by benjamin-work** [comment]
Technically, your example shows how to unfreeze parameters but the issue is about how to freeze layers ;)

Therefore, if you intend to add this to the docs/notebook, I would also like to explicitly see parameters being frozen.

**Comment by thomasjpfan** [comment]
What about this:

```python
from skorch.helper import filtered_optimizer

module = module_cls()
net = Net(module, optimizer=filtered_optimizer(torch.optim.SGD))

# Freeze first layer
for p in module.dense0.parameters():
    p.requires_grad_(False)

X, y = data
net.fit(X, y)

# Unfreeze first layer
for p in module.dense0.parameters():
    p.requires_grad_(True)

net.initialize_optimizer()
net.partial_fit(X, y)
```

**Comment by taketwo** [comment]
My two cents:
 * `filtered_optimizer` is not a clear function name. True, it reflects the operation that is happening inside it, but does not tell what is being filtered, which is arguably more important. 
 * add helpers for freezing/unfreezing?

**Comment by benjamin-work** [comment]
> filtered_optimizer is not a clear function name

True, do you have other suggestions?

> add helpers for freezing/unfreezing?

May not be necessary, since we should strive to introduce as little abstraction as possible to pytorch. Furthermore, if we want to have helpers, we can always add them later.

**Comment by thomasjpfan** [comment]
Some options are: 

1. `filter_requires_grad`
2. `filter_frozen_parameters`
3. `gen_optimizer_filtering_requires_grad`

**Comment by taketwo** [comment]
I think `filter_requires_grad()` is already better.

Though stepping back a little, wouldn't this be a cleaner syntax:

```python
NeuralNet(..., optimizer=torch.optim.SGD, optimizer__filter_requires_grad=True)
```

`filter_requires_grad` is a "virtual" argument of course, we won't pass it to the optimizer constructor, but instead apply the decorator.

**Comment by thomasjpfan** [comment]
What does everyone think of adding the parameter `optimizer__filter_requires_grad` to `NeutralNet`?

**Comment by benjamin-work** [comment]
> What does everyone think of adding a parameter similar to optimizer__filter_requires_grad to NeutralNet?

As far as I see, that would set a precedent, which really doesn't seem worth it. Again, I wish that this was supported by pytorch directly ...

**Comment by taketwo** [comment]
One way or another, in the end we are modifying the behavior of PyTorch optimizers. It's just the syntax that is different:

```python
net = NeuralNet(..., optimizer=torch.optim.SGD, optimizer__filter_requires_grad=True)
```

vs.

```python
net = NeuralNet(..., optimizer=optimizer_params_requires_grad(torch.optim.SGD))
```

From my perspective, the former is more expressive. Also, if freezing/unfreezing is a variable in your script, it's so easy to directly pass it as parameter.

**Comment by BenjaminBossan** [comment]
Still 1. sets a precedent and 2. doesn't, and making the latter depend on a parameter is only a one-liner.

**Comment by benjamin-work** [comment]
I will update the `Advanced_Usage` notebook with an example of freezing and unfreezing tomorrow. From my perspective, there are still numerous ways to do it wrong, so there is still a need for further simplification, but at least it is possible without too much hassle.

**Comment by thomasjpfan** [comment]
I am planning to convert this pytorch tutorial: https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html, so users can compare using skorch and pure pytorch. In that case, they are just freezing everything but the final layer.

**Comment by benjamin-work** [comment]
That would be a great addition. I believe there is no harm in having two places where freezing is explained.

**Comment by benjamin-work** [comment]
Just FYI, I have been working on this but am not satisfied with the outcome. In my opinion, there are too many steps involved (and skipping some can still *appear* to work but not with the intended result):

1. Tell the optimizer to exclude parameters that don't require gradients.
2. Change the requires_grad attribute of parameters to be frozen/unfrozen.
3. Re-initialize the optimizer.
4. Call `partial_fit` on the new data.

Explaining why this is needed requires too much knowledge of skorch internals for my taste. Things would be easier if `NeuralNet(..., module__linear0__weight__requires_grad=False)` were possible (see #107, #273), making 2. easier and 3. and 4. obsolete if I see it right. Therefore, I would postpone that PR until we have a solution for recursively setting params (which itself is blocked because of the difficulty of setting params on sub-modules).

For the time being, we have #281 which hopefully already covers most use cases.

**Comment by ottonemo** [comment]
For release 0.3.0 this is sufficient but we should keep this in mind for later releases.

**Comment by benjamin-work** [comment]
Since #273 is stalling and freezing/unfreezing is still too complicated and error prone IMO, I was thinking about a callback that could help with that, along the lines of:

```
class SetRequiresGradParams(Callback):
    def __init__(self, match_func, requires_grad=False):
        self.match_func = match_func
        self.requires_grad = requires_grad
        
    def on_train_begin(self, net, **kwargs):
        for name, param in net.module_.named_parameters():
            if self.match_func(name):
                param.requires_grad_(self.requires_grad)
```

The `match_func` can be any callable (`re.match`, `fnmatch`, `operator.eq`, ...) and this callback can be used both for freezing and unfreezing. Still not as nice as `module__weight0__requires_grad=False` but better than the status quo.

What do you all think?

**Comment by thomasjpfan** [comment]
I like the `SetRequiresGradParams` idea. 👍

**Comment by ottonemo** [comment]
With the addition of `Freezer` we now have documentation in form of example notebooks and an easy way to freeze layers.
